### PR TITLE
service: add new RestartMode option

### LINKED ISSUE #20412: Service enters "failed" state even if it is restarted
I would like to distinguish between failed services that will be restarted from ones that will not be restarted anymore.

Currently, when active service fails it enters "failed" state before entering "activating" again due to Restart= entry in the service file. This makes it hard to determine whether the service has failed but will be restarted or it has failed and will not be restarted.

I was reading the following documentation and it suggests that service enters “failed” state only after it can’t be restarted anymore:
https://www.freedesktop.org/software/systemd/man/systemd.service.html
> Note that service restart is subject to unit start rate limiting configured with StartLimitIntervalSec= and StartLimitBurst=, see systemd.unit(5) for details. **A restarted service enters the failed state only after the start limits are reached.**

Either it's a bug or the sentence is not precise. This is an example of a service that fails three times before it stops being restarted due to StartLimitBurst=. State changes were reported through dbus org.freedesktop.DBus.Properties.PropertiesChanged signal on Unit.

```
test.service:
[Unit]
StartLimitIntervalSec=10
StartLimitBurst=3

[Service]
Type=simple
ExecStart=bash -c "sleep 1; exit 1"
Restart=on-failure
```
```
busctl monitor --system org.freedesktop.systemd1
systemctl start test.service
```
```
ActiveState   SubState
--------------------------
inactive      dead
active        running
failed        failed
activating    auto-restart
inactive      dead
active        running
failed        failed
activating    auto-restart
inactive      dead
active        running
failed        failed
activating    auto-restart
inactive      dead
failed        failed
```
As you can see it enters failed state multiple times so it's not possible to know which one is final.

Suggestions:
1) Don't enter "failed" state if the service will be restarted, the service can still recover so it has not failed yet. Go from active straight to activating:
```
ActiveState   SubState
--------------------------
active        running
(failure)
activating    auto-restart
```

2) Enter "inactive" state instead of "failed" if the service will be restarted
```
ActiveState   SubState
--------------------------
active        running
inactive      dead
activating    auto-restart
```

3) Specify in substate that the service will be restarted, for example:
```
ActiveState   SubState
--------------------------
active        running
failed        will-restart
activating    auto-restart
```

`systemd 247 (247.3-6)`

### PR DESCRIPTION
**Pull request by rphibel**
When this option is true, the service restarts without entering a failed state. Dependent units are not notified of transitory failure.
    
This is useful for the following use case:
    

> We have a target with `Requires=my-service`, `After=my-service`. `my-service.service` is a oneshot service and has `Restart=on-failure` in its definition.
> 
> `my-service` can get stuck for various reasons and time out, in which case it is restarted. Currently, when it fails the first time, the target fails, even though `my-service` is restarted.
> 
> The behavior we're looking for is that until `my-service` is not restarted anymore, the target stays pending waiting for `my-service` to start successfully or fail without being restarted anymore.

Closes #20412.

### CODE REVIEW

File: man/systemd.service.xml [hunk 1]
```diff
@@ -898,6 +898,15 @@
         </listitem>
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>RestartQuick=</varname></term>
```
Comment (yuwata) [inline man/systemd.service.xml]: "quick" sounds ambiguous. Please make the option more descriptive.
Maybe `NotifyFailureOnRestart=` is better and defaults to `yes` ?

File: test/units/testsuite-03.sh [hunk 2]
```diff
@@ -110,4 +110,10 @@ ELAPSED=$((END_SEC-START_SEC))
 [[ "$ELAPSED" -ge 3 ]] && [[ "$ELAPSED" -le 5 ]] || exit 1
 [[ "$RESULT" -ne 0 ]] || exit 1
 
+systemctl start succeeds-on-restart.target || true
```
Comment (yuwata) [inline test/units/testsuite-03.sh]: Why `|| true` is added?

File: test/units/testsuite-03.sh [hunk 3]
```diff
@@ -110,4 +110,10 @@ ELAPSED=$((END_SEC-START_SEC))
...
+assert_rc 0 systemctl --quiet is-active succeeds-on-restart.target
+
+systemctl start fails-on-restart.target || true
```
Comment (yuwata) [inline test/units/testsuite-03.sh]: Please use `|| :`.

**Review by yuwata — CHANGES_REQUESTED** [review]
Superficial comments.

**Comment by rphibel** [comment]
> What's the use case for this?

This is the followup of this PR: https://github.com/systemd/systemd/pull/26751

In some cases, we don't want a dependent unit to fail if a dependency fails **temporarily**, i.e. fails at first but succeeds after restart. The use case is described in PR: https://github.com/systemd/systemd/pull/26751:

> We have a target with Requires=my-service, After=my-service. my-service.service is a oneshot service and has Restart=on-failure in its definition.
> 
> my-service.service can get stuck for various reasons and time out, in which case it is restarted. Currently, when it fails the first time, the target fails, even though my-service is restarted.
> 
> The behavior we're looking for is that until my-service is not restarted anymore, the target stays pending waiting for my-service.service to start successfully or fail without being restarted anymore.

File: man/systemd.service.xml (hunk 1 above)
Comment (rphibel) [reply man/systemd.service.xml]: This is the name that was suggested in PR https://github.com/systemd/systemd/pull/26751.

@poettering @DaanDeMeyer Are you ok with `NotifyFailureOnRestart=` instead of `RestartQuick=`?

File: test/units/testsuite-03.sh [hunk 4]
```diff
@@ -110,4 +110,10 @@ ELAPSED=$((END_SEC-START_SEC))
 [[ "$ELAPSED" -ge 3 ]] && [[ "$ELAPSED" -le 5 ]] || exit 1
 [[ "$RESULT" -ne 0 ]] || exit 1
 
+systemctl start succeeds-on-restart.target
+assert_rc 0 systemctl --quiet is-active succeeds-on-restart.target
+
+systemctl start fails-on-restart.target || :
+assert_rc 3 systemctl --quiet is-active fails-on-restart.target
```
Comment (yuwata) [inline test/units/testsuite-03.sh]: Could you also add testcases for the default behavior, that is, RestartQuick=no (or NotifyFailureOnRestart=yes) ?

File: test/units/testsuite-03.sh (hunk 3 above)
Comment (rllane17) [reply test/units/testsuite-03.sh]: - - - [ ] # ~~@

 ~~

File: man/systemd.service.xml [hunk 5]
```diff
@@ -898,6 +898,18 @@
...
```
Comment (poettering) [inline man/systemd.service.xml]: hmm, i sense we might need to extend this later, hence maybe call this RestartMode= or so, and make it an enum, which takes "normal" or "quick" as values?

File: man/systemd.service.xml [hunk 6]
```diff
@@ -898,6 +898,18 @@
...
+
+        <listitem><para>Takes a boolean value that specifies whether the service should restart quick,
+        i.e. without going through a failed state. Defaults to <option>no</option>.</para>
```
Comment (poettering) [inline man/systemd.service.xml]: this should make clear what is and what isn't skipped. i.e. say whether ExecStop= and such are invoked, and what about OnSuccess= and OnFailure=

File: man/systemd.service.xml (hunk 5 above)
Comment (poettering) [reply man/systemd.service.xml]: (background: i am not entirely sure going through "deactivating" state is really always desirable, i.e. running ExecStop= and friends. hence we might need to extend this later)

**Review by poettering — CHANGES_REQUESTED** [review]

**Comment by dtardon** [comment]
This resolves #20412.

File: man/systemd.service.xml [hunk 7]
```diff
@@ -898,6 +898,26 @@
...
       </varlistentry>
 
+      <varlistentry>
+        <term><varname>RestartMode=</varname></term>
+
+        <listitem>
+          <para>Takes a string value that specifies how the service should restart:
+            <itemizedlist>
+              <listitem><para>If set to <option>normal</option> (the default), the service restarts by
+              going through a failed state.</para></listitem>
```
Comment (YHNdnzj) [inline man/systemd.service.xml]: A service can also enter inactive state rather than failed if `Restart=always` is used.

File: test/units/testsuite-03.sh [hunk 8]
```diff
@@ -3,6 +3,9 @@
 set -eux
 set -o pipefail
 
+# shellcheck source=test/units/assert.sh
+. "$(dirname "$0")"/assert.sh
```
Comment (yuwata) [inline test/units/testsuite-03.sh]: This is recently renamed to util.sh.

File: man/systemd.service.xml (hunk 7 above)
Comment (rphibel) [reply man/systemd.service.xml]: I did a test with the following service file:
```
[Service]
Type=simple
ExecStart=false
Restart=always
```
The service does enter a failed state (`failed-before-auto-restart` to be more precise):
```
May 21 19:47:56 final systemd[1]: test-always.service: Changed dead -> running
May 21 19:47:56 final systemd[1]: test-always.service: Changed running -> failed-before-auto-restart
May 21 19:47:56 final systemd[1]: test-always.service: Changed failed-before-auto-restart -> auto-restart
May 21 19:47:56 final systemd[1]: test-always.service: Changed auto-restart -> dead
```

Comment (YHNdnzj) [reply man/systemd.service.xml]: That's because the service is actually failing (since you're calling `false`). You can probably refer to logs in https://github.com/systemd/systemd/issues/27722 for more information (or just change it to `true`).

Comment (daandemeyer) [reply man/systemd.service.xml]: Yeah if you change this to execute `true` instead of `false` it will go through dead-before-auto-restart instead of failed-before-auto-restart which is inactive instead of failed

File: man/systemd.service.xml [hunk 9]
```diff
@@ -897,6 +897,28 @@
...
+
+        <listitem>
+          <para>Takes a string value that specifies how a failing service should restart:
+            <itemizedlist>
+              <listitem><para>If set to <option>normal</option> (the default), the service restarts by
+              going through a failed state.</para></listitem>
+
+              <listitem><para>If set to <option>quick</option>, the service restarts without
+              going through a failed state.
+              <varname>ExecStopPost=</varname> and <varname>OnSuccess=</varname> are invoked.
```
Comment (daandemeyer) [inline man/systemd.service.xml]: I wonder if we want to invoke `OnSuccess=` in  this scenario.

**Review by DaanDeMeyer — APPROVED** [review]
Generally looks good to me but I'd also want @poettering's stamp of approval

File: man/systemd.service.xml (hunk 9 above)
Comment (YHNdnzj) [reply man/systemd.service.xml]: I'd say we shouldn't. And `OnSuccess=` is actually getting triggered as a side-effect of #27722. If that's fixed, with this patch `OnSuccess=` will be skipped too.

**Comment by rphibel** [comment]
@poettering Did you get a chance to look at this one?

**Comment by YHNdnzj** [comment]
@rphibel As mentioned, please drop the statement regarding `OnSuccess=`. That should be considered as a bug, and shall not be documented in that way.

**Comment by poettering** [comment]
hmm, i am a bit puzzled by the docs, since it only talks about failing services. as i read the code this also affects services with Restart=always which exit cleanly, i.e. are not failure. Because for them OnSuccess= is also skipped. 

Also, the docs claim ExecStop= wasn't called if the new mode is used. But afaics it totally is, the state engine is not changed in that regard. 

/me is confused now?

**Comment by rphibel** [comment]
> hmm, i am a bit puzzled by the docs, since it only talks about failing services. as i read the code this also affects services with Restart=always which exit cleanly, i.e. are not failure. Because for them OnSuccess= is also skipped.
> 
> Also, the docs claim ExecStop= wasn't called if the new mode is used. But afaics it totally is, the state engine is not changed in that regard.
> 
> /me is confused now?

True, services exiting cleanly are also affected and the behaviour of ExecStop= is not changed. I corrected the doc

File: man/systemd.service.xml [hunk 10]
```diff
@@ -897,6 +897,28 @@
...
+          <para>Takes a string value that specifies how a service should restart:
+            <itemizedlist>
+              <listitem><para>If set to <option>normal</option> (the default), the service restarts by
+              going through a failed state.</para></listitem>
+
+              <listitem><para>If set to <option>quick</option>, the service restarts without
+              going through a failed state.
+              <varname>ExecStopPost=</varname> is invoked.
+              <varname>OnFailure=</varname> is skipped.</para></listitem>
+            </itemizedlist>
```
Comment (poettering) [inline man/systemd.service.xml:913]: this focusses only on the "failed" state. But your patch also disables the "inactive" state we go through between the shutting down and starting up. 

And because we enter that inactive state we also will run OnActive=. With your new option this will *also* be turned off.

**Review by poettering — CHANGES_REQUESTED** [review]

File: man/systemd.service.xml [hunk 11]
```diff
@@ -897,6 +897,28 @@
...
+        <listitem>
+          <para>Takes a string value that specifies how a service should restart:
+            <itemizedlist>
+              <listitem><para>If set to <option>normal</option> (the default), the service restarts by
+              going through a failed/inactive state.</para></listitem>
+
+              <listitem><para>If set to <option>quick</option>, the service restarts without
+              going through a failed/inactive state.
+              <varname>ExecStopPost=</varname> is invoked.
+              <varname>OnFailure=</varname> is skipped.</para></listitem>
```
Comment (poettering) [inline man/systemd.service.xml]: OnSuccess= is skipped too, also deserves mention.

Comment (rphibel) [reply man/systemd.service.xml]: Actually `OnSuccess=` is called when the service goes from the `auto-restart` state to the `dead` state

Comment (poettering) [reply man/systemd.service.xml]: hmm?

Since a1d315730ffddf283d4bb9d73878fbcd97a4d244 we definitely go into low-level state     SERVICE_DEAD_BEFORE_AUTO_RESTART (which maps to high-level UNIT_DEAD) before entering auto-restart. This should trigger the OnSuccess= dependency whenever we terminate a service successfully and then restart.

Comment (rphibel) [reply man/systemd.service.xml]: Yes, this is what I see. Here are the states we go through for a service exiting cleanly and with `Restart=always`
- `RestartMode=normal` : `running -> stop -> stop-post -> dead-before-auto-restart -> auto-restart -> dead -> running`
- `RestartMode=quick` : `running -> stop -> stop-post -> auto-restart -> dead -> running`

For a failing service:
- `RestartMode=normal` : `running -> stop-post -> failed-before-auto-restart -> auto-restart -> dead -> running`
- `RestartMode=quick` : `running -> stop-post -> auto-restart -> dead -> running`

Comment (YHNdnzj) [reply man/systemd.service.xml]: Yeah, that's https://github.com/systemd/systemd/issues/27722.

Comment (daandemeyer) [reply man/systemd.service.xml]: So let's mention here that `OnSuccess=` is skipped. Yes it is still called due to #27722 but that's a bug that shouldn't be taken into account here.

**Review by DaanDeMeyer — APPROVED** [review]
LGTM, but let's mention that OnSuccess= is skipped, as the only reason that it is triggered is because of another bug

File: man/systemd.service.xml [hunk 12]
```diff
@@ -897,6 +897,28 @@
...
+        <term><varname>RestartMode=</varname></term>
+
+        <listitem>
+          <para>Takes a string value that specifies how a service should restart:
+            <itemizedlist>
+              <listitem><para>If set to <option>normal</option> (the default), the service restarts by
+              going through a failed/inactive state.</para></listitem>
+
+              <listitem><para>If set to <option>quick</option>, the service restarts without
+              going through a failed/inactive state.
```
Comment (YHNdnzj) [inline man/systemd.service.xml]: "If set to quick, the service transitions to the activating state directly during auto-restart, skipping failed/inactive state."

Maybe something like this is clearer?

**Review by YHNdnzj — APPROVED** [review]

**Comment by DaanDeMeyer** [comment]
This is still missing the OnSuccess= mention

File: man/systemd.service.xml (hunk 11 above)
Comment (YHNdnzj) [reply man/systemd.service.xml]: And that's already fixed by #28215.

**Comment by poettering** [comment]
Could you rebase and fix the OnSuccess= thing, now that this is properly fixed via #28215? We'd like to release v254-rc1 soon, and would be good if we could include this one in the upcoming release.

**Comment by poettering** [comment]
lgtm

**Comment by YHNdnzj** [comment]
@rphibel I had just rebased and added missing SPDX for you, and now they got removed again...

**Comment by poettering** [comment]
@YHNdnzj just rebase and push?

File: man/systemd.service.xml [hunk 13]
```diff
@@ -897,6 +897,28 @@
...
+      <varlistentry>
+        <term><varname>RestartMode=</varname></term>
+
+        <listitem>
+          <para>Takes a string value that specifies how a service should restart:
+            <itemizedlist>
+              <listitem><para>If set to <option>normal</option> (the default), the service restarts by
+              going through a failed/inactive state.</para></listitem>
+
+              <listitem><para>If set to <option>quick</option>, the service transitions to the activating
```
Comment (bluca) [inline man/systemd.service.xml]: bike shedding, but: `quick` seems awfully unclear to me, as this is not about speed, but skipping some steps. So how about `skip-notifications` or `skip-triggers` instead?

Comment (poettering) [reply man/systemd.service.xml]: dunno, it's about skipping a state. the notifications/triggers are just an effect of the skipped state.

still find "quick" much more descriptive.

also the word "trigger" we use for .timer, .socket, .path units. I'd really not overload this here again.

and notifications is the wrong word too. we act on a dep, which is not precisely the same as a notification.

hence, still think "quick" is a good name, and better than the proposed alternatives.

File: test/testsuite-03.units/succeeds-on-restart.sh [hunk 14]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
```
Comment (bluca) [inline test/testsuite-03.units/succeeds-on-restart.sh:1]: SPDX lines are missing here and from other new files

File: man/systemd.service.xml (hunk 13 above)
Comment (bluca) [reply man/systemd.service.xml]: `skip-inactive-state`? It's how it's described in this manpage - `quick` by itself really doesn't convey the right message, ie, the opposite wouldn't be `slow`

Comment (poettering) [reply man/systemd.service.xml]: i don't like skip-inactive-state, since it skips both the inactive state and the failed state.

Comment (bluca) [reply man/systemd.service.xml]: By popular acclamation (a telegram poll) we have settled on `direct` instead of `quick` - @rphibel could you please do the change? Then it's good to go

**Review by bluca — APPROVED** [review]

File: src/core/job.c [hunk 15]
```diff
@@ -235,7 +235,7 @@ Job* job_install(Job *j) {
                         /* not conflicting, i.e. mergeable */
 
                         if (uj->state == JOB_WAITING ||
-                            (job_type_allows_late_merge(j->type) && job_type_is_superset(uj->type, j->type))) {
+                            (job_type_allows_late_merge(j->type) && mode != JOB_RESTART_DEPENDENCIES && job_type_is_superset(uj->type, j->type))) {
```
Comment (YHNdnzj) [inline src/core/job.c:238]: This is new, and not related to the addition of `RestartMode=`. Can you do this in a separate commit with explanations?

**Review by YHNdnzj — COMMENTED** [review]
"quick" in the commit message is not updated.

File: src/core/transaction.c [hunk 16]
```diff
@@ -657,7 +657,7 @@ static int transaction_apply(
                 /* Clean the job dependencies */
                 transaction_unlink_job(tr, j, false);
 
-                installed_job = job_install(j);
+                installed_job = job_install(j, mode);
```
Comment (YHNdnzj) [inline src/core/transaction.c:660]: This change belongs to the wrong commit?

Comment (rphibel) [reply src/core/transaction.c:660]: Sorry I missed this one
